File: c/producer.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <assert.h>
#include <stdatomic.h>
#include <stddef.h>
#include <stdbool.h>

#include "spsc.c"

#define PRODUCER_CONCAT3(a, b, c) a##b##c
#define PRODUCER_EXPAND_THEN_CONCAT3(a, b, c) PRODUCER_CONCAT3(a, b, c)
#define PRODUCER_(x) PRODUCER_EXPAND_THEN_CONCAT3(CHANNEL_NAME, _producer_, x)

typedef struct PRODUCER_(producer) {
    QUE_(spsc_t) * spsc;                  /* Pointer to shared SPSC */
    uint64_t tail;                        /* Local tail */
    uint64_t written;                     /* Number of written elements since last sync */
    uint64_t last_consumer_heartbeat;     /* Last known consumer heartbeat */
} PRODUCER_(producer_t);

/* Reservation struct for batch writes */
typedef struct PRODUCER_(reservation) {
    PRODUCER_(producer_t) *producer;
    uint64_t start_tail;
    uint64_t count;
    uint64_t written;
    bool committed;
} PRODUCER_(reservation_t);

static inline int
PRODUCER_(join_or_initialize)( void *shmem_region, PRODUCER_(producer_t)* producer ) {

    /* Capacity must be a power of two */
    assert( CHANNEL_N > 0 && (CHANNEL_N & (CHANNEL_N - 1)) == 0 );  

    /* Ensure alignment */
    assert( ((size_t)shmem_region % ALIGNMENT) == 0 ); 

    QUE_(spsc_t) *spsc = (QUE_(spsc_t) *)shmem_region;

    /* Check magic */
    uint64_t magic = spsc->magic;
    uint64_t capacity = spsc->capacity;

    /* Join existing queue */
    if( magic == MAGIC ) {
        if( capacity != CHANNEL_N ) {
            fprintf( stderr, "Incorrect capacity\n" );
            return 1;
        }

        /* Increment producer heartbeat to signal we've joined */
        atomic_fetch_add_explicit( &(spsc->producer_heartbeat.value), 1, memory_order_release );

        producer->spsc = spsc;
        producer->tail = atomic_load_explicit( &(spsc->tail.value), memory_order_acquire );
        producer->written = 0;
        producer->last_consumer_heartbeat = atomic_load_explicit( &(spsc->consumer_heartbeat.value),
                                                                  memory_order_acquire );
        return 0;
    }
    
    /* Initialization */
    if( magic == 0 ) {
        /* Initialize the SPSC structure */
        atomic_store_explicit( &(spsc->tail.value), 0, memory_order_release );
        atomic_store_explicit( &(spsc->head.value), 0, memory_order_release );
        atomic_store_explicit( &(spsc->producer_heartbeat.value), 0, memory_order_release );
        atomic_store_explicit( &(spsc->consumer_heartbeat.value), 0, memory_order_release );
        spsc->capacity = CHANNEL_N;
        spsc->magic = MAGIC;

        producer->spsc = spsc;
        producer->tail = 0;
        producer->written = 0;
        producer->last_consumer_heartbeat = atomic_load_explicit( &(spsc->consumer_heartbeat.value), 
                                                                  memory_order_acquire );
        return 0;
    }

    fprintf( stderr, "Corruption detected\n" );
    return 1;
}

/* Join existing initialized channel */
static inline int
PRODUCER_(join)( void *shmem_region, PRODUCER_(producer_t)* producer ) {

    /* Capacity must be a power of two */
    assert( CHANNEL_N > 0 && (CHANNEL_N & (CHANNEL_N - 1)) == 0 );  

    /* Ensure alignment */
    assert( ((size_t)shmem_region % ALIGNMENT) == 0 ); 

    QUE_(spsc_t) *spsc = (QUE_(spsc_t) *)shmem_region;

    uint64_t magic = spsc->magic;
    uint64_t capacity = spsc->capacity;
    
    if( magic == MAGIC ) {
        if( capacity != CHANNEL_N ) {
            fprintf( stderr, "Incorrect capacity\n" );
            return 1;
        }

        producer->spsc = spsc;
        producer->tail = atomic_load_explicit( &(spsc->tail.value), memory_order_acquire );
        producer->written = 0;
        producer->last_consumer_heartbeat = atomic_load_explicit( &(spsc->consumer_heartbeat.value),
                                                                  memory_order_acquire );
        return 0;
    }
    
    if( magic == 0 ) {
        fprintf( stderr, "Uninitialized channel\n" );
        return 1;
    }

    fprintf( stderr, "Corruption detected\n" );
    return 1;
}

/* Synchronize the local tail with shared memory */
static inline void
PRODUCER_(sync)( PRODUCER_(producer_t) *producer ) {
    producer->written = 0;
    atomic_store_explicit( &(producer->spsc->tail.value), producer->tail, memory_order_release );
}

/* Push with backpressure - returns 0 on success, 1 if full */
static inline int
PRODUCER_(push)( PRODUCER_(producer_t) *producer, void *value ) {
    QUE_(spsc_t) *spsc = producer->spsc;

    /* Check if queue is full */
    uint64_t head = atomic_load_explicit( &(spsc->head.value), memory_order_acquire );
    int is_full = (producer->tail == head + CHANNEL_N);

    if( is_full ) {
        return 1; /* Queue is full */
    }

    /* Calculate the index in the circular buffer */
    uint64_t index = producer->tail & (CHANNEL_N - 1);

    /* Get buffer pointer (starts at offset 640) */
    void *buffer_base = spsc + 1;
    void *addr = (char *)buffer_base + (index * sizeof(CHANNEL_T));

    /* Write the value to the buffer */
    memcpy( addr, value, sizeof(CHANNEL_T) );

    /* Increment tail and written counter */
    producer->tail++;
    producer->written++;

    // /* Sync if we've written enough */
    // if( producer->written == burst_amount(CHANNEL_N) ) {
    //     PRODUCER_(sync)(producer);
    // }

    return 0; /* Success */
}

static inline void
PRODUCER_(beat)( PRODUCER_(producer_t) *producer ) {
    QUE_(spsc_t) *spsc = producer->spsc;

    /* Atomically increment the producer's heartbeat */
    atomic_fetch_add_explicit(&(spsc->producer_heartbeat.value), 1, memory_order_release);
}

static inline int 
PRODUCER_(consumer_heartbeat)( PRODUCER_(producer_t) *producer ) {
    QUE_(spsc_t) *spsc = producer->spsc;

    /* Load the current heartbeat of the consumer */
    uint64_t heartbeat = atomic_load_explicit( &(spsc->consumer_heartbeat.value), memory_order_acquire );

    /* Check if the consumer's heartbeat has changed */
    if( heartbeat != producer->last_consumer_heartbeat ) {
        /* Update the last known consumer heartbeat */
        producer->last_consumer_heartbeat = heartbeat;
        return 1; /* Consumer is active */
    }

    return 0; /* Consumer has not updated its heartbeat */
}

/* Get pointer to padding area (can be used for metadata) */
static inline void*
PRODUCER_(get_padding_ptr)( PRODUCER_(producer_t) *producer ) {
    return producer->spsc->padding;
}

/* Reserve space for batch writes */
static inline int
PRODUCER_(reserve)( PRODUCER_(producer_t) *producer, uint64_t count, PRODUCER_(reservation_t) *reservation ) {
    if( count == 0 || count > CHANNEL_N ) {
        return 1; /* Invalid size */
    }

    /* Check if we have space */
    uint64_t head = atomic_load_explicit( &(producer->spsc->head.value), memory_order_acquire );
    uint64_t available_space = CHANNEL_N - (producer->tail - head);
    
    if( count > available_space ) {
        return 1; /* Not enough space */
    }

    /* Initialize reservation */
    reservation->producer = producer;
    reservation->start_tail = producer->tail;
    reservation->count = count;
    reservation->written = 0;
    reservation->committed = false;

    return 0; /* Success */
}

/* Write next value in a reservation */
static inline CHANNEL_T *
PRODUCER_(reservation_get_next)( PRODUCER_(reservation_t) *reservation ) {
    // assert( reservation->written < reservation->count );

    uint64_t index = (reservation->start_tail + reservation->written) & (CHANNEL_N - 1);
    
    /* Get buffer pointer */
    void *buffer_base = reservation->producer->spsc + 1;
    void *addr = (char *)buffer_base + (index * sizeof(CHANNEL_T));

    return addr;
}

/* Write next value in a reservation */
static inline void
PRODUCER_(reservation_write_next)( PRODUCER_(reservation_t) *reservation ) {
    reservation->written++;
}

/* Write all values from array to reservation */
static inline void
PRODUCER_(reservation_write_all)( PRODUCER_(reservation_t) *reservation, void *values, uint64_t count ) {
    // assert( reservation->written + count <= reservation->count );

    if( count == 0 ) {
        return;
    }

    uint64_t start_index = (reservation->start_tail + reservation->written) & (CHANNEL_N - 1);
    uint64_t end_index = start_index + count;
    
    void *buffer_base = reservation->producer->spsc + 1;

    if( end_index <= CHANNEL_N ) {
        /* No wraparound - single copy */
        memcpy( (char *)buffer_base + (start_index * sizeof(CHANNEL_T)),
                values,
                count * sizeof(CHANNEL_T) );
    } else {
        /* Wraparound - two copies */
        uint64_t first_part_len = CHANNEL_N - start_index;
        
        /* Copy first part (until end of buffer) */
        memcpy( (char *)buffer_base + (start_index * sizeof(CHANNEL_T)),
                values,
                first_part_len * sizeof(CHANNEL_T) );
        
        /* Copy second part (from beginning of buffer) */
        memcpy( buffer_base,
                (char *)values + (first_part_len * sizeof(CHANNEL_T)),
                (count - first_part_len) * sizeof(CHANNEL_T) );
    }

    reservation->written += count;
}

/* Commit a reservation */
static inline void
PRODUCER_(reservation_commit)( PRODUCER_(reservation_t) *reservation ) {
    reservation->committed = true;

    /* Only advance tail by amount actually written */
    reservation->producer->tail += reservation->written;
    reservation->producer->written += reservation->written;

    /* Sync  */
    PRODUCER_(sync)(reservation->producer);
}

/* Cancel a reservation without publishing */
static inline void
PRODUCER_(reservation_cancel)( PRODUCER_(reservation_t) *reservation ) {
    /* Nothing to do - just don't advance tail */
    reservation->committed = true; /* Mark as handled */
}

File: que/src/lossless/producer.rs
use std::{ptr::NonNull, sync::Arc};

use bytemuck::AnyBitPattern;

use crate::{
    atomic_compat::Ordering,
    error::QueError, page_size::PageSize, shmem::Shmem, ChannelMode,
    ShmemMode, MAGIC,
};

use super::{burst_amount, Channel};

#[repr(C, align(128))]
pub struct Producer<M: ChannelMode<T>, T, const N: usize> {
    spsc: NonNull<Channel<M, T, N>>,
    tail: usize,
    head_cache: usize,
    /// Number of elements written since last sync
    written: usize,
    last_consumer_heartbeat: usize,
}

unsafe impl<M: ChannelMode<T>, T, const N: usize> Send
    for Producer<M, T, N>
{
}

impl<T: AnyBitPattern, const N: usize> Producer<ShmemMode, T, N> {
    /// Joins or creates a channel backed by shared memory as a
    /// producer.
    pub unsafe fn join_or_create_shmem(
        shmem_id: &str,
        #[cfg(target_os = "linux")] page_size: PageSize,
    ) -> Result<Producer<ShmemMode, T, N>, QueError> {
        #[cfg(not(target_os = "linux"))]
        let page_size = PageSize::Standard;

        // Calculate buffer size.
        // If using huge pages, we must uplign to page size.
        let buffer_size: i64 = page_size
            .mem_size(core::mem::size_of::<Channel<ShmemMode, T, N>>())
            .try_into()
            .map_err(|_| QueError::InvalidSize)?;

        // Open or create shmem
        let shmem = Shmem::open_or_create(
            shmem_id,
            buffer_size,
            #[cfg(target_os = "linux")]
            page_size,
        )?;

        // Zerocopy deserialize the SPSC
        let spsc: *mut Channel<ShmemMode, T, N> =
            shmem.get_mut_ptr().cast();

        // Check magic
        let magic = (*spsc).magic.load(Ordering::Acquire);
        let capacity = (*spsc).capacity.load(Ordering::Acquire);
        #[rustfmt::skip]
        return if magic == MAGIC {
            // Check capacity
            if capacity != N {
                return Err(QueError::IncorrectCapacity(capacity));
            }

            // Successful join if magic and capacity is correct
            Ok(Producer {
                spsc: NonNull::new(shmem.get_mut_ptr().cast()).unwrap(),
                tail: (*spsc).tail.load(Ordering::Acquire),
                head_cache: (*spsc).head.load(Ordering::Acquire),
                written: 0,
                last_consumer_heartbeat: (*spsc)
                    .consumer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else if magic == 0 {
            (*spsc).tail.store(0, Ordering::Release);
            (*spsc).producer_heartbeat.store(0, Ordering::Release);
            (*spsc).consumer_heartbeat.store(0, Ordering::Release);
            (*spsc).capacity.store(N, Ordering::Release);
            (*spsc).magic.store(MAGIC, Ordering::Release);

            Ok(Producer {
                spsc: NonNull::new(shmem.get_mut_ptr().cast()).unwrap(),
                tail: 0,
                head_cache: 0,
                written: 0,
                last_consumer_heartbeat: (*spsc)
                    .consumer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else {
            // Magic is not MAGIC and not zero
            Err(QueError::CorruptionDetected)
        };
    }

    /// Initializes a channel backed by `buffer` and joins as a
    /// producer.
    ///
    /// SAFETY:
    /// This must point to a buffer of proper size and alignment.
    pub unsafe fn join_or_initialize_in(
        buffer: *mut u8,
    ) -> Result<Producer<ShmemMode, T, N>, QueError> {
        Self::join_or_initialize_in_(buffer)
    }

    /// Joins an existing channel backed by `buffer` as a producer.
    ///
    /// SAFETY:
    /// This must point to a buffer of proper size and alignment.
    pub unsafe fn join(
        buffer: *mut u8,
    ) -> Result<Producer<ShmemMode, T, N>, QueError> {
        Self::join_(buffer)
    }
}

impl<M: ChannelMode<T>, T, const N: usize> Producer<M, T, N> {
    pub const MODULO_MASK: usize = N - 1;

    pub(crate) unsafe fn join_or_initialize_in_(
        buffer: *mut u8,
    ) -> Result<Producer<M, T, N>, QueError> {
        assert!(
            N > 0 && N.is_power_of_two(),
            "Capacity must be a power of two"
        );
        assert!(buffer as usize % 128 == 0, "unaligned");

        // Zerocopy deserialize the SPSC
        let spsc: *mut Channel<M, T, N> = buffer.cast();

        // Check magic
        let magic = (*spsc).magic.load(Ordering::Acquire);
        let capacity = (*spsc).capacity.load(Ordering::Acquire);
        #[rustfmt::skip]
        return if magic == MAGIC {
            // Check capacity
            if capacity != N {
                return Err(QueError::IncorrectCapacity(capacity));
            }

            (*spsc)
                .producer_heartbeat
                .fetch_add(1, Ordering::Release);

            if M::BACKED_BY_ARCC {
                unsafe {
                    Arc::increment_strong_count(spsc);
                }
            }

            // Successful join if magic and capacity is correct
            Ok(Producer {
                spsc: NonNull::new(buffer.cast()).unwrap(),
                tail: (*spsc).tail.load(Ordering::Acquire),
                head_cache: (*spsc).head.load(Ordering::Acquire),
                written: 0,
                last_consumer_heartbeat: (*spsc)
                    .consumer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else if magic == 0 {
            (*spsc).tail.store(0, Ordering::Release);
            (*spsc).consumer_heartbeat.store(0, Ordering::Release);
            (*spsc).producer_heartbeat.store(0, Ordering::Release);
            (*spsc).capacity.store(N, Ordering::Release);
            (*spsc).magic.store(MAGIC, Ordering::Release);

            if M::BACKED_BY_ARCC {
                unsafe {
                    Arc::increment_strong_count(spsc);
                }
            }

            Ok(Producer {
                spsc: NonNull::new(buffer.cast()).unwrap(),
                tail: 0,
                head_cache: 0,
                written: 0,
                last_consumer_heartbeat: (*spsc)
                    .consumer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else {
            // Magic is not MAGIC and not zero
            Err(QueError::CorruptionDetected)
        };
    }

    pub(crate) unsafe fn join_(
        buffer: *mut u8,
    ) -> Result<Producer<M, T, N>, QueError> {
        assert!(
            N > 0 && N.is_power_of_two(),
            "Capacity must be a power of two"
        );
        assert!(buffer as usize % 128 == 0, "unaligned");

        // Zerocopy deserialize the SPSC
        let spsc: *mut Channel<M, T, N> = buffer.cast();

        let magic = (*spsc).magic.load(Ordering::Acquire);
        let capacity = (*spsc).capacity.load(Ordering::Acquire);
        if magic == MAGIC {
            if capacity != N {
                return Err(QueError::IncorrectCapacity(capacity));
            }

            if M::BACKED_BY_ARCC {
                unsafe {
                    Arc::increment_strong_count(spsc);
                }
            }

            // Successful join if magic and capacity is correct
            Ok(Producer {
                spsc: NonNull::new(buffer.cast()).unwrap(),
                tail: (*spsc).tail.load(Ordering::Acquire),
                head_cache: (*spsc).head.load(Ordering::Acquire),
                written: 0,
                last_consumer_heartbeat: (*spsc)
                    .consumer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else if magic == 0 {
            // Technically could be corrupted but uninitialized
            // is most likely explanation
            Err(QueError::Uninitialized)
        } else {
            // Magic is not MAGIC and not zero
            Err(QueError::CorruptionDetected)
        }
    }

    /// Attempts to write a new element to the channel. If full, returns
    /// [QueError::Full].
    #[inline(always)]
    pub fn push(&mut self, value: T) -> Result<(), QueError> {
        // Check if full
        if self.tail == self.head_cache + N {
            self.head_cache = unsafe {
                (*self.spsc.as_ptr())
                    .head
                    .load(Ordering::Acquire)
            };
            if self.tail == self.head_cache + N {
                return Err(QueError::Full);
            }
        }

        // Write value if not full
        let index = self.tail & Self::MODULO_MASK;
        unsafe {
            *(*self.spsc.as_ptr())
                .buffer
                .as_mut_ptr()
                .add(index) = value;
        };

        // Increment tail and written counter
        self.tail += 1;
        self.written += 1;

        if self.written == burst_amount::<N>() {
            self.sync();
        }

        Ok(())
    }

    /// Increments the producer heartbeat.
    ///
    /// Can be read by the consumer to see that the producer is still
    /// online if done periodically. Can also be used to ack individual
    /// messages or alert that we've joined.
    pub fn beat(&self) {
        unsafe {
            (*self.spsc.as_ptr())
                .producer_heartbeat
                .fetch_add(1, Ordering::Release);
        }
    }

    /// Synchronizes the local tail with the atomic tail in the channel,
    /// publishing newly written values.
    #[inline(always)]
    pub fn sync(&mut self) {
        self.written = 0;
        unsafe {
            (*self.spsc.as_ptr())
                .tail
                .store(self.tail, Ordering::Release)
        }
    }

    /// Checks if a consumer has incremented its heartbeat since last
    /// called. Can be used by the producer to see if the consumer is
    /// still online if done periodically. Can also be used to ack
    /// individual messages or alert that we've joined.
    pub fn consumer_heartbeat(&mut self) -> bool {
        let heartbeat = unsafe {
            (*self.spsc.as_ptr())
                .consumer_heartbeat
                .load(Ordering::Acquire)
        };

        if heartbeat != self.last_consumer_heartbeat {
            self.last_consumer_heartbeat = heartbeat;
            true
        } else {
            false
        }
    }

    /// Returns pointer to inner padding.
    ///
    /// User is responsible for safe usage.
    ///
    /// Can be used to store metadata (e.g. hash seed).
    ///
    /// Byte array is 128 byte aligned.
    pub fn get_padding_ptr(&self) -> NonNull<[u8; 112]> {
        unsafe {
            NonNull::new_unchecked(
                self.spsc.cast::<u8>().as_ptr().add(512),
            )
            .cast()
        }
    }

    /// Reserves space for multiple elements. Returns a reservation that must be
    /// committed to publish the values.
    #[inline(always)]
    pub fn reserve(
        &mut self,
        count: usize,
    ) -> Result<Reservation<'_, M, T, N>, QueError> {
        if count == 0 {
            return Err(QueError::InvalidSize);
        }

        if count > N {
            return Err(QueError::InvalidSize);
        }

        let mut available_space = N - (self.tail - self.head_cache);
        if count > available_space {
            self.head_cache = unsafe {
                (*self.spsc.as_ptr())
                    .head
                    .load(Ordering::Acquire)
            };
            available_space = N - (self.tail - self.head_cache);
            if count > available_space {
                return Err(QueError::Full);
            }
        }

        Ok(Reservation {
            start_tail: self.tail,
            producer: self,
            count,
            written: 0,
            committed: false,
        })
    }
}

/// A reservation of space in the queue that can be written to
pub struct Reservation<'a, M: ChannelMode<T>, T, const N: usize> {
    producer: &'a mut Producer<M, T, N>,
    start_tail: usize,
    count: usize,
    written: usize,
    committed: bool,
}

impl<'a, M: ChannelMode<T>, T, const N: usize>
    Reservation<'a, M, T, N>
{
    /// Returns the number of slots reserved
    #[inline(always)]
    pub fn len(&self) -> usize {
        self.count
    }

    /// Returns the number of values written so far
    #[inline(always)]
    pub fn written(&self) -> usize {
        self.written
    }

    /// Returns the number of slots remaining to be written
    #[inline(always)]
    pub fn remaining(&self) -> usize {
        self.count - self.written
    }

    /// Write the next value in sequence
    ///
    /// # Panics
    /// Panics if all reserved slots have been written
    #[inline(always)]
    pub fn write_next(&mut self, value: T) {
        if self.written >= self.count {
            panic!("Attempted to write beyond reservation limit: written {} of {} slots", 
                   self.written, self.count);
        }

        let index = (self.start_tail + self.written)
            & Producer::<M, T, N>::MODULO_MASK;
        unsafe {
            *(*self.producer.spsc.as_ptr())
                .buffer
                .as_mut_ptr()
                .add(index) = value;
        }

        self.written += 1;
    }

    /// Write all values from a slice
    ///
    /// # Panics
    /// Panics if the slice is larger than the remaining space in the reservation
    #[inline(always)]
    pub fn write_all(&mut self, values: &[T])
    where
        T: Copy,
    {
        let remaining = self.remaining();
        if values.len() > remaining {
            panic!("Attempted to write {} values with only {} slots remaining", 
                   values.len(), remaining);
        }

        if values.is_empty() {
            return;
        }

        let start_index = (self.start_tail + self.written)
            & Producer::<M, T, N>::MODULO_MASK;
        let end_index = start_index + values.len();

        unsafe {
            let buffer_ptr = (*self.producer.spsc.as_ptr())
                .buffer
                .as_mut_ptr();

            if end_index <= N {
                // No wraparound - single copy
                std::ptr::copy_nonoverlapping(
                    values.as_ptr(),
                    buffer_ptr.add(start_index),
                    values.len(),
                );
            } else {
                // Wraparound - two copies
                let first_part_len = N - start_index;

                // Copy first part (until end of buffer)
                std::ptr::copy_nonoverlapping(
                    values.as_ptr(),
                    buffer_ptr.add(start_index),
                    first_part_len,
                );

                // Copy second part (from beginning of buffer)
                std::ptr::copy_nonoverlapping(
                    values.as_ptr().add(first_part_len),
                    buffer_ptr,
                    values.len() - first_part_len,
                );
            }
        }

        self.written += values.len();
    }

    /// Write values from an iterator
    ///
    /// Stops when the reservation is full or the iterator is exhausted.
    /// Returns the number of values written.
    #[inline(always)]
    pub fn write_iter<I>(&mut self, iter: &mut I) -> usize
    where
        I: Iterator<Item = T>,
    {
        let start_written = self.written;

        for value in iter {
            if self.written >= self.count {
                break;
            }
            self.write_next(value);
        }

        self.written - start_written
    }

    /// Commit the reservation, publishing all written values
    ///
    /// If not all reserved slots were written, only the written values are published.
    #[inline(always)]
    pub fn commit(mut self) {
        self.committed = true;

        // Only advance tail by the amount actually written
        self.producer.tail += self.written;
        self.producer.written += self.written;

        // Sync if we've written enough
        if self.producer.written >= burst_amount::<N>() {
            self.producer.sync();
        }
    }

    /// Cancel the reservation without publishing any values
    #[inline(always)]
    pub fn cancel(self) {
        /* cancel is actually a noop lol */
    }
}

impl<M: ChannelMode<T>, T, const N: usize> Drop for Producer<M, T, N> {
    fn drop(&mut self) {
        // LocalMode is backed by arc
        if M::BACKED_BY_ARCC {
            unsafe { drop(Arc::from_raw(self.spsc.as_ptr())) }
        }
    }
}
